# STAT 207 Group Lab Assignment 13 - [10 total points]

## Measuring Model Performance & Model Searching

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

Grade this submission.

If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  

Also, discuss and record if you've ever been on a cruise before.


None of us have been on a cruise.

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to assess model fit and overfitting through model searching.
<hr>

### Group Roles

Suggested and specified roles are provided below: 

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

### Imports

In [46]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as smf
from sklearn.model_selection import train_test_split

## Case Study: Crew Size on Cruise Ships

We will look at data collected about the size of a crew (in hundreds) on cruise ships.  In particular, we would like to be able to predict the size of the crew needed for a cruise ship from other characteristics about that ship.  The original data can be found here: https://github.com/bot13956/ML_Model_for_Predicting_Ships_Crew_Size/blob/master/cruise_ship_info.csv.

For this case study, we will consider the following variables:

- **Age**: the age of the ship (years)
- **Tonnage**: the amount of water the ship displaces
- **passengers**: the max number of passengers the ship can hold (in 100s)
- **length**: the length of the ship (in hundrds of feet)
- **cabins**: the number of cabins (in 100s)
- **passenger_density**: a space ratio for how much space is on board per passenger
- **crew**: the number of crew members on the ship (in 100s)

The code cell below will read in the data for you and create a test and training set.  Be sure to run the cell. 

In [47]:
df = pd.read_csv('cruise_ship_info.csv')
df_train, df_test = train_test_split(df, test_size = 0.2, random_state = 425)

### 1. [4 points] Searching for a Good Model

**a)** Perform forward selection using the adjusted $R^2$ ($R^2_\text{adj}$) as your metric on your training data.

In [48]:
current_model = smf.ols('crew ~ 1', data=df_train).fit()
current_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   crew   R-squared:                      -0.000
Model:                            OLS   Adj. R-squared:                 -0.000
Method:                 Least Squares   F-statistic:                       nan
Date:                Wed, 23 Apr 2025   Prob (F-statistic):                nan
Time:                        15:50:33   Log-Likelihood:                -338.48
No. Observations:                 126   AIC:                             679.0
Df Residuals:                     125   BIC:                             681.8
Df Model:                           0                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      7.8694      0.318     24.772      0.000       7.241       8.498
==============================================================================
Omnibus:                        5.753   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.056   Jarque-Bera (JB):                6.234
Skew:                           0.309   Prob(JB):                       0.0443
Kurtosis:                       3.898   Cond. No.                         1.00
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [49]:
variables = ['Age', 'Tonnage', 'passengers', 'length', 'cabins', 'passenger_density']

for var in variables:
    formula = f'crew ~ {var}'
    model = smf.ols(formula, data=df_train).fit()
    print(f'{var}: Adjusted R² = {model.rsquared_adj:.5f}')

Age: Adjusted R² = 0.34191
Tonnage: Adjusted R² = 0.86783
passengers: Adjusted R² = 0.85316
length: Adjusted R² = 0.79757
cabins: Adjusted R² = 0.92011
passenger_density: Adjusted R² = -0.00258


In [50]:
current_model = smf.ols('crew ~ cabins', data=df_train).fit()
variables = ['Age', 'Tonnage', 'passengers', 'length', 'passenger_density']

for var in variables:
    formula = f'crew ~ cabins + {var}'
    model = smf.ols(formula, data=df_train).fit()
    print(f'{var}: Adjusted R² = {model.rsquared_adj:.5f}')

Age: Adjusted R² = 0.92075
Tonnage: Adjusted R² = 0.92331
passengers: Adjusted R² = 0.92226
length: Adjusted R² = 0.92712
passenger_density: Adjusted R² = 0.92534


In [51]:
current_model = smf.ols('crew ~ cabins + length', data=df_train).fit()
variables = ['Age', 'Tonnage', 'passengers', 'passenger_density']

for var in variables:
    formula = f'crew ~ cabins + length + {var}'
    model = smf.ols(formula, data=df_train).fit()
    print(f'{var}: Adjusted R² = {model.rsquared_adj:.5f}')

Age: Adjusted R² = 0.92694
Tonnage: Adjusted R² = 0.92685
passengers: Adjusted R² = 0.93101
passenger_density: Adjusted R² = 0.92923


In [52]:
current_model = smf.ols('crew ~ cabins + length + passengers', data=df_train).fit()
variables = ['Age', 'Tonnage', 'passenger_density']

for var in variables:
    formula = f'crew ~ cabins + length + passengers + {var}'
    model = smf.ols(formula, data=df_train).fit()
    print(f'{var}: Adjusted R² = {model.rsquared_adj:.5f}')

Age: Adjusted R² = 0.93123
Tonnage: Adjusted R² = 0.93213
passenger_density: Adjusted R² = 0.93150


In [53]:
current_model = smf.ols('crew ~ cabins + length + passengers + Tonnage', data=df_train).fit()
variables = ['Age', 'passenger_density']

for var in variables:
    formula = f'crew ~ cabins + length + passengers + Tonnage + {var}'
    model = smf.ols(formula, data=df_train).fit()
    print(f'{var}: Adjusted R² = {model.rsquared_adj:.5f}')

Age: Adjusted R² = 0.93175
passenger_density: Adjusted R² = 0.93159


In [54]:
forward_model = smf.ols('crew ~ cabins + length + passengers + Tonnage', data=df_train).fit()
forward_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   crew   R-squared:                       0.934
Model:                            OLS   Adj. R-squared:                  0.932
Method:                 Least Squares   F-statistic:                     430.2
Date:                Wed, 23 Apr 2025   Prob (F-statistic):           1.67e-70
Time:                        15:50:34   Log-Likelihood:                -166.95
No. Observations:                 126   AIC:                             343.9
Df Residuals:                     121   BIC:                             358.1
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.5085      0.630     -0.807      0.421      -1.756       0.740
cabins         0.7955      0.089      8.974      0.000       0.620       0.971
length         0.3137      0.120      2.607      0.010       0.075       0.552
passengers    -0.1306      0.040     -3.237      0.002      -0.211      -0.051
Tonnage        0.0152      0.009      1.735      0.085      -0.002       0.032
==============================================================================
Omnibus:                      132.153   Durbin-Watson:                   2.125
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3664.597
Skew:                           3.452   Prob(JB):                         0.00
Kurtosis:                      28.502   Cond. No.                         664.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**b)** Briefly summarize the results of this selection process.  Be sure to answer:
- What was your starting model?
- What was the most important initial variable to add to the model?
- What was your final model?
- What was the $R^2_\text{adj}$ of your final model?

The starting model was 'crew ~ 1', so just the intercept. The most important initial variable to add to the model was cabins. It led to the largest jump in the model's adjusted R-squared. The final model was 'crew ~ cabins + length + passengers + Tonnage' and its R-squared was 0.932.

### 2. [2.5 points]  Check for Multicollinearity

**a)** Now, check your model that you selected in Question 1 after forward selection for multicollinearity.  You may use either a visualization or summary statistics to check.

In [55]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from patsy import dmatrices

y, X = dmatrices('crew ~ cabins + length + passengers + Tonnage', data=df_train, return_type='dataframe')

vif_df = pd.DataFrame()
vif_df['Variable'] = X.columns
vif_df['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

print(vif_df)

     Variable        VIF
0   Intercept  58.019096
1      cabins  23.666153
2      length   6.770778
3  passengers  23.224946
4     Tonnage  16.451396


**b)** Based on what you see from part a, which variables (if any) would you like to remove from your model to reduce multicollinearity?  Briefly explain how you selected your variables that you wanted to remove.  Then, fit a new model to the training data using your reduced set of selected predictors.

*Note*: If you are satisfied with your current model for multicollinearity, pick the least important predictor to remove from the model, and fit an updated model to the training data with your selected predictors.  Be sure to explain why you did not want to remove any variables.

*Second note*: there is not a single correct answer here.  We are hoping that you make a reasonable choice based on your current understanding of the data.

I removed cabins because it had the highest collinearity. Then I removed Tonnage because the VIFs were still over 10 in the reduced model. At that point, the collinearity under 10 for all variables so I left it at that.

In [56]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from patsy import dmatrices

y, X = dmatrices('crew ~ length + passengers + Tonnage', data=df_train, return_type='dataframe')

vif_df = pd.DataFrame()
vif_df['Variable'] = X.columns
vif_df['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

print(vif_df)

     Variable        VIF
0   Intercept  57.974577
1      length   6.699548
2  passengers  10.545761
3     Tonnage  15.457914


In [57]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
from patsy import dmatrices

y, X = dmatrices('crew ~ length + passengers', data=df_train, return_type='dataframe')

vif_df = pd.DataFrame()
vif_df['Variable'] = X.columns
vif_df['VIF'] = [variance_inflation_factor(X.values, i) for i in range(X.shape[1])]

print(vif_df)

     Variable        VIF
0   Intercept  42.751955
1      length   4.557747
2  passengers   4.557747


In [58]:
reduced_model = smf.ols('crew ~ length + passengers', data=df_train).fit()
reduced_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   crew   R-squared:                       0.882
Model:                            OLS   Adj. R-squared:                  0.880
Method:                 Least Squares   F-statistic:                     457.9
Date:                Wed, 23 Apr 2025   Prob (F-statistic):           1.03e-57
Time:                        15:50:34   Log-Likelihood:                -204.06
No. Observations:                 126   AIC:                             414.1
Df Residuals:                     123   BIC:                             422.6
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -1.9778      0.720     -2.745      0.007      -3.404      -0.552
length         0.6997      0.131      5.322      0.000       0.439       0.960
passengers     0.2202      0.024      9.253      0.000       0.173       0.267
==============================================================================
Omnibus:                       72.073   Durbin-Watson:                   1.945
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1057.278
Skew:                           1.504   Prob(JB):                    2.60e-230
Kurtosis:                      16.869   Cond. No.                         150.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 3. [3.5 points] Comparing Our Two Models for Overfitting

We now have two candidate models: our model from forward selection in Question 1 and our reduced model from Question 2.  In this question, we'll aim to identify which model performs better on *new* data.

**a)** Calculate the RMSE for each of our two models on each of our two data sets: the training and the test set.  Be sure that you use the same original model for both calculations (that is, you don't need to re-fit a second version of your model to your test data).

In [59]:
from sklearn.metrics import mean_squared_error
import numpy as np

y_train_actual = df_train['crew']
y_train_pred_fwd = forward_model.predict(df_train)
y_train_pred_reduced = reduced_model.predict(df_train)

rmse_train_fwd = np.sqrt(mean_squared_error(y_train_actual, y_train_pred_fwd))
rmse_train_reduced = np.sqrt(mean_squared_error(y_train_actual, y_train_pred_reduced))


In [60]:
y_test_actual = df_test['crew']
y_test_pred_fwd = forward_model.predict(df_test)
y_test_pred_reduced = reduced_model.predict(df_test)

rmse_test_fwd = np.sqrt(mean_squared_error(y_test_actual, y_test_pred_fwd))
rmse_test_reduced = np.sqrt(mean_squared_error(y_test_actual, y_test_pred_reduced))

In [61]:
print(f"Forward selected Model - Training RMSE: {rmse_train_fwd:.3f}")
print(f"Forward selected Model - Testing RMSE:  {rmse_test_fwd:.3f}")
print(f"Reduced Model - Training RMSE: {rmse_train_reduced:.3f}")
print(f"Reduced Model - Testing RMSE:  {rmse_test_reduced:.3f}")

Forward selected Model - Training RMSE: 0.910
Forward selected Model - Testing RMSE:  1.168
Reduced Model - Training RMSE: 1.222
Reduced Model - Testing RMSE:  1.355


**b)** Based on these results, which model would you prefer?  Briefly explain how you picked your preferred model.

Although my full model has high collinearity, it has higher accuracy on both the training and testing sets. For this reason, I would prefer the full, forward-selected model.

**c)** Do your results from Question 1 (forward searching) and Question 3 (RMSE) suggest using the same model?  

What characteristic of a model is prioritized by each of the metrics used in Questions 1 ($R^2_\text{adj})$, 2 (multicollinearity), and 3 (RMSE)?  

*Note:* You should have one answer for each of the three metrics.

Questions 1 and 3 suggest using the full, forward selected model. R-squared prioritizes parsimoniousness, multicollinearity prioritizes stability, and RMSE prioritizes pure accuracy.